### ISSUE TITLE
Undo after using the number scrubber or color picker takes more steps than expected

### ISSUE THREAD

**Issue by pamelafox** [Khan/live-editor#273]
To replicate:

1. Write `<img id="hopper" src="https://www.kasandbox.org/programming-images/creatures/Hopper-Happy.png" width="88">` in a webpage
2. Scrub 88 up to 150.
3. Use CTRL-Z to undo. Notice how many steps it takes you to get back to 88 (a lot).

I think this is the sort of thing we can be happy to live with, but am noting here since it has been reported.

**Comment by MikaalSky** [comment]
I think this should get fixed. Something like don't save the state until user takes mouse off scrubber.

**Comment by kevinb7** [comment]
I had a quick look at the public API for the UndoManager in Ace and there are no hooks for controlling what actually goes on to the undo/redo stack.  It would definitely be nice to have this but It will require modifying Ace which is something I'd only consider if those changes can be merged back into Ace.  They already do undo merging when typing.  Providing hooks that allow for undo merging in user defined scenarios may not be out of the question.

**Comment by kevinb7** [comment]
https://github.com/ajaxorg/ace/issues/2369

**Comment by ChrisJPhoenix** [comment]
It looks like you might just reach in from js/ui/tooltips/number-scrubber.js and tweak the boolean parent.editor.session.$fromUndo variable to prevent undo's from being saved at all. It's an undocumented interface, but I think it'll do what you want and not do anything you don't want. (The point of the variable is not to save changes resulting from hitting "undo" so although its purpose is different its effect should be what we want.)

Is this too hacky, or shall I develop the code and do a pull request?

**Comment by kevinb7** [comment]
@ChrisJPhoenix it doesn't sound like a huge change.  Please include tests so that if we update ace and it breaks this we'll know.  Thanks for looking into this.

**Comment by ChrisJPhoenix** [comment]
I thought I might get it done before move-day, but no... it'll probably have to be early next week. 

How do I invoke the test suite?

**Comment by kevinb7** [comment]
No worries.  As for the tests there's a couple of ways they can be run.  If you run ```gulp``` it will start a watcher and as you make changes to source files it will compile those changes and then run the appropriate test suite against the build files.  You can run all of the tests using ```gulp test```, unfortunately I just noticed "test_output_js" is failing right now because the tests are running inside an iframe and gulp-mocha-phantomjs can't handle that.  I need to fix this.  

It's probably easiest to run the tests in the browser.  To do this you'll to make sure that the all of the files in the repo are being served by a static server.  This can be accomplished by running ```python -m SimpleHTTPServer```.  Then open ```localhost:8000``` in a browser and navigate to each of the ```index.html``` files in the tests/ folder (don't worry about tests/output/debugging).

**Comment by kevinb7** [comment]
@ChrisJPhoenix Also, before we can accept a pull request from you, you'll need to sign the CLA.
www.khanacademy.org/r/cla.

**Comment by ChrisJPhoenix** [comment]
I've signed the CLA. 

To write the test you requested, I want to test a sub-sequence of multiple undo events. But I can't find a way to keep Ace from merging deltas when I enter text. I've tried several things - text with newlines (google says at one point Ace wouldn't merge deltas across newlines), and moving the cursor to insert text in the middle of previously-inserted text - do you know a straightforward way?

**Comment by kevinb7** [comment]
Does your code merge deltas when using the scrubber?  Try to simulate that code path.  Could you post a link to the code you have so far or create a pull request so that people can look at?  It's much easier to give advice when looking at code. :)

**Comment by kevinb7** [comment]
Thanks for signing the CLA and thanks for looking into this bug.  This will be a big win for usability.

**Comment by ChrisJPhoenix** [comment]
OK, I think I've successfully created a pull request. My code doesn't merge deltas; that would need lots of hacking in ACE data structures. I just tell ACE that the incoming changes are coming from the "Undo" key and shouldn't be remembered. This has some problems; see comments; I don't know if this approach will be workable.

**Comment by kevinb7** [comment]
Thanks for the pull request.  I left some comments on https://github.com/Khan/live-editor/pull/347.  It's good to link to the pull request from the issue (or vice verse) so that it's easy to find.  Also, you may want to use branches for future pull requests.  That way you can grab the latest upstream changes without having to merge first.

**Comment by kevinb7** [comment]
Fixed in https://github.com/Khan/live-editor/commit/49b40b9185ec458b58fb1bd48f3cf927e4458f24 thanks to @ChrisJPhoenix.  It works like a charm.  I'm sure that all of our users will be super happy to see this new feature.  It hasn't been deployed yet.  I'll do that either this evening or tomorrow.
